File: bindings/cpp/test/test_transaction.cpp
// Transactions — remote-only (contract §6 "Transactions").
#include "harness.hpp"
#include "kevy/client.hpp"
#include "kevy/transaction.hpp"

using namespace kevy;
using namespace kevy::test;

KEVY_TEST(txn_multi_exec_order) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());
  Transaction tx = c.multi();
  tx.set("a", "1").incr("n").get("a");
  auto replies = tx.exec();
  CHECK_EQ(replies.size(), size_t(3));
  CHECK(replies[0].kind == ReplyKind::Simple && replies[0].str() == "OK");
  CHECK(replies[1].kind == ReplyKind::Int && replies[1].integer == 1);
  CHECK(replies[2].kind == ReplyKind::Bulk && replies[2].str() == "1");
}

KEVY_TEST(txn_typed_cursor) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());
  c.set("k", "v");  // commit before the txn (kevy EXEC has no intra-txn read-your-writes)
  Transaction tx = c.multi();
  tx.set("other", "x").incr("ctr").mget({"k"});
  TransactionReplies cur = tx.exec_typed();
  CHECK_EQ(cur.remaining(), size_t(3));
  cur.next_ok();
  CHECK_EQ(cur.next_int(), int64_t(1));
  auto mg = cur.next_array_of_bulks();
  CHECK_EQ(mg.size(), size_t(1));
  CHECK(mg[0].has_value() && *mg[0] == "v");
  cur.expect_empty();  // arity gate: all consumed
}

KEVY_TEST(txn_watch_abort) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());
  Client other = Client::connect(srv->url());
  c.set("wk", "0");
  c.watch({"wk"});
  Transaction tx = c.multi();
  tx.set("wk", "in-txn");
  other.set("wk", "changed");  // concurrent modify by another client
  auto res = tx.exec_watched();
  CHECK(!res.has_value());  // WATCH violation → abort (null)
}

KEVY_TEST(txn_watch_success) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());
  c.set("wk", "0");
  c.watch({"wk"});
  Transaction tx = c.multi();
  tx.set("wk", "committed");
  auto res = tx.exec_watched();  // no concurrent modify → commit
  CHECK(res.has_value());
  CHECK_EQ(res->size(), size_t(1));
  auto v = c.get("wk");
  CHECK(v.has_value() && *v == "committed");
}

KEVY_TEST(txn_abandon_implicit_discard) {
  if (!server_available()) return;
  auto srv = spawn_server();
  Client c = Client::connect(srv->url());
  {
    Transaction tx = c.multi();
    tx.set("a", "queued");
    // tx goes out of scope without exec/discard → RAII implicit DISCARD.
  }
  // The socket must not be stuck in MULTI: a normal command still works.
  c.set("ok", "yes");
  auto v = c.get("ok");
  CHECK(v.has_value() && *v == "yes");
  // The abandoned SET must not have applied.
  CHECK(!c.get("a").has_value());
}

File: bindings/python/tests/test_transaction.py
"""Transactions: MULTI / EXEC / DISCARD + WATCH (contract §3.12, §6).
Remote-only."""

import kevy
from kevy import ReplyKind


def test_multi_queue_exec_order(remote_server):
    c = kevy.connect(remote_server.url)
    tx = c.multi()
    tx.set("a", "1").incr("n").get("a")
    replies = tx.exec()
    assert len(replies) == 3
    assert replies[0].kind is ReplyKind.SIMPLE and replies[0].data == b"OK"
    assert replies[1].integer == 1
    assert replies[2].data == b"1"
    c.close()


def test_exec_typed_cursor(remote_server):
    c = kevy.connect(remote_server.url)
    c.set("pre", "p")  # committed before the block (see deviation note below)
    tx = c.multi()
    tx.set("a", "1").incr("n").mget("pre", "missing")
    cur = tx.exec_typed()
    cur.next_ok()
    assert cur.next_int() == 1
    assert cur.next_array_of_bulks() == [b"p", None]
    cur.expect_empty()  # arity gate
    c.close()


def test_mget_in_multi_sees_earlier_write(remote_server):
    # Read-your-writes inside a MULTI: an MGET queued after a SET of the same
    # key must observe that write, exactly like GET does. This was a server
    # deviation (the cross-shard gather raced ahead of the still-buffered
    # single-key write); fixed server-side — see client-contract.md §3.12
    # (EXEC = array of N typed replies, one per queued command, executed in
    # order).
    c = kevy.connect(remote_server.url)
    tx = c.multi()
    tx.set("w", "1").mget("w")
    replies = tx.exec()
    assert replies[1].items[0].data == b"1"
    c.close()


def test_watch_abort(remote_server):
    c = kevy.connect(remote_server.url)
    other = kevy.connect(remote_server.url)
    c.watch("wk")
    other.set("wk", "changed")  # concurrent modify after WATCH
    tx = c.multi()
    tx.set("wk", "mine")
    assert tx.exec_watched() is None  # WATCH violation → abort
    assert other.get("wk") == b"changed"
    c.close()
    other.close()


def test_abandon_sends_implicit_discard(remote_server):
    c = kevy.connect(remote_server.url)
    with c.multi() as tx:
        tx.set("x", "1")
        # leave the block without exec/discard — __exit__ sends DISCARD
    # socket is not stuck in MULTI mode: a normal command works
    c.set("y", "ok")
    assert c.get("y") == b"ok"
    assert c.get("x") is None  # the queued SET was discarded
    c.close()


def test_queue_typed_builders(remote_server):
    c = kevy.connect(remote_server.url)
    tx = c.multi()
    tx.queue(b"SET", b"k", b"v")
    replies = tx.exec()
    assert replies[0].data == b"OK"
    c.close()
